use PlaceRef abstractions more consistently

henryboisdequin opened the pull request:
Addresses this [comment](https://github.com/rust-lang/rust/pull/80865/files#r558978715)
Associated issue: #80647

r? @RalfJung

henryboisdequin commented:
@rustbot label: +C-cleanup +A-mir +T-compiler

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -962,8 +962,7 @@ impl<'tcx> LocalDecl<'tcx> {
                     pat_span: _,
-                })
-                | BindingForm::ImplicitSelf(ImplicitSelfKind::Imm),
+                }) | BindingForm::ImplicitSelf(ImplicitSelfKind::Imm),
```
What are these changes about?

henryboisdequin replied:
Sorry, just my formatter, will change that

RalfJung commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -1025,13 +1025,14 @@ fn place_as_reborrow(
-                    ty::Ref(..) => Some(place_base.projection),
-                    _ => None,
+
+                if let ty::Ref(..) = inner_ty.kind() {
```
```suggestion
                if matches!(inner_ty.kind(), ty::Ref(..)) =  {
```

RalfJung commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -492,7 +492,7 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
         match *rvalue {
             Rvalue::Ref(_, kind, place) => {
-                if let Some(reborrowed_proj) = place_as_reborrow(self.tcx, self.body, place) {
+                if let Some(place_ref) = place_as_reborrow(self.tcx, self.body, place) {
```
Something involving "reborrow" in the same would still be a good idea, I think.

RalfJung commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -508,20 +508,20 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                         }
                     };
                     self.visit_local(&place.local, ctx, location);
```
This should now probably use `place_ref.local`.

RalfJung commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -508,20 +508,20 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                     };
                     self.visit_local(&place.local, ctx, location);
-                    self.visit_projection(place.local, reborrowed_proj, ctx, location);
+                    self.visit_projection(place.local, place_ref.projection, ctx, location);
```
Same here, `place_ref.local`.

Really `Visitor` should be changed such that `` takes a `PlaceRef` instead of taking the local and the projections separately.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1897356351) [(plain)](https://github.com/rust-lang/rust/commit/9c575bf4ee7be64ac849e54d24446bb3a1e3691e/checks/1897356351/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
error: missing condition for `if` expression
    --> compiler/rustc_mir/src/transform/check_consts/validation.rs:1055:19
     |
1055 |                 if matches!(inner_ty.kind(), ty::Ref(..)) =  {
     |                   ^ expected if condition here
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error[E0308]: mismatched types
    --> compiler/rustc_mir/src/transform/check_consts/validation.rs:1055:20
     |
     |
1055 |                   if matches!(inner_ty.kind(), ty::Ref(..)) =  {
1056 | |                     return Some(place_base);
1057 | |                 } else {
1057 | |                 } else {
     | |_________________^ expected `bool`, found `()`
     |
help: you might have meant to use pattern matching
     |
1055 |                 if let matches!(inner_ty.kind(), ty::Ref(..)) =  {
     |                    ^^^
help: you might have meant to compare for equality
     |
1055 |                 if matches!(inner_ty.kind(), ty::Ref(..)) ==  {

error: unreachable expression
    --> compiler/rustc_mir/src/transform/check_consts/validation.rs:1055:20
     |
     |
1055 |                   if matches!(inner_ty.kind(), ty::Ref(..)) =  {
1056 | |                     return Some(place_base);
     | |                     ----------------------- any code following this expression is unreachable
1057 | |                 } else {
     | |_________________^ unreachable expression
     | |_________________^ unreachable expression
     |
     = note: `-D unreachable-code` implied by `-D warnings`
error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_mir`
error: could not compile `rustc_mir`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_interface" "-p" "rustc_resolve" "-p" "rustc_index" "-p" "rustc_macros" "-p" "rustc_arena" "-p" "rustc_privacy" "-p" "rustc_ast_passes" "-p" "rustc_attr" "-p" "rustc_lexer" "-p" "rustc_builtin_macros" "-p" "rustc_parse_format" "-p" "rustc_typeck" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_ty_utils" "-p" "rustc_traits" "-p" "rustc_expand" "-p" "rustc_lint_defs" "-p" "rustc_passes" "-p" "rustc_ast_lowering" "-p" "rustc_mir_build" "-p" "rustc_apfloat" "-p" "rustc_symbol_mangling" "-p" "rustc_trait_selection" "-p" "rustc_incremental" "-p" "rustc_fs_util" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_session" "-p" "rustc_span" "-p" "rustc_plugin_impl" "-p" "rustc_ast" "-p" "rustc_middle" "-p" "rustc_type_ir" "-p" "rustc_query_system" "-p" "rustc_lint" "-p" "rustc_feature" "-p" "rustc_hir_pretty" "-p" "rustc_errors" "-p" "rustc_parse" "-p" "rustc_error_codes" "-p" "rustc_serialize" "-p" "rustc_metadata" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_target" "-p" "rustc_save_analysis" "-p" "rustc_ast_pretty" "-p" "rustc_data_structures" "-p" "rustc_hir" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:16
```

</details>

RalfJung commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -1029,7 +1029,7 @@ fn place_as_reborrow(
                 let inner_ty = place_base.ty(body, tcx).ty;
 
-                if let ty::Ref(..) = inner_ty.kind() {
+                if matches!(inner_ty.kind(), ty::Ref(..)) =  {
```
*oops* I forgot to also remove the `=`

henryboisdequin commented:
@RalfJung Update:

- Addressed all of your comments
- I'm currently changing `visit_projection` and all of its associated functions to take a `PlaceRef`

henryboisdequin commented:
@rustbot label: +S-waiting-on-author -S-waiting-on-review

henryboisdequin commented:
@RalfJung Nearly finished but I still have some errors and don't know how to resolve them. The current code is pushed, as you can see. Here are the errors:

```
error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:154:34
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
...    |
154  | |                 self.super_place(place_ref, context, location);
     | |                                  ^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mut mir::PlaceRef<'tcx>`
...    |
929  | |     }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1084 |   make_mir_visitor!(Visitor,);
     |   ---------------------------- in this macro invocation

error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:383:29
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
...    |
383  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
...    |
929  | |     }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1084 |   make_mir_visitor!(Visitor,);
     |   ---------------------------- in this macro invocation
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`

error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:390:29
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
...    |
390  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
...    |
929  | |     }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1084 |   make_mir_visitor!(Visitor,);
     |   ---------------------------- in this macro invocation
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`

error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:383:29
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
...    |
383  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
...    |
929  | |     }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1085 |   make_mir_visitor!(MutVisitor, mut);
     |   ----------------------------------- in this macro invocation
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`

error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:390:29
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
...    |
390  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
...    |
929  | |     }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1085 |   make_mir_visitor!(MutVisitor, mut);
     |   ----------------------------------- in this macro invocation
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle`
```

I'm not really sure why they are happening or how to fix them. Help would be appreciated!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1903187930) [(plain)](https://github.com/rust-lang/rust/commit/226106667d03c7c31925fb745283c5febbc83936/checks/1903187930/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:154:34
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
...    |
154  | |                 self.super_place(place_ref, context, location);
     | |                                  ^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mut mir::PlaceRef<'tcx>`
929  | |     }
930  | | }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1084 |   make_mir_visitor!(Visitor,);

error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:383:29
     |
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
383  | |                             &mut place.as_ref(),
383  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
929  | |     }
930  | | }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1084 |   make_mir_visitor!(Visitor,);
     |
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`
error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:390:29
     |
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
390  | |                             &mut place.as_ref(),
390  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
929  | |     }
930  | | }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1084 |   make_mir_visitor!(Visitor,);
     |
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`
error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:383:29
     |
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
383  | |                             &mut place.as_ref(),
383  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
929  | |     }
930  | | }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1085 |   make_mir_visitor!(MutVisitor, mut);
     |
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`
error[E0308]: mismatched types
    --> compiler/rustc_middle/src/mir/visit.rs:390:29
     |
     |
68   | / macro_rules! make_mir_visitor {
69   | |     ($visitor_trait_name:ident, $($mutability:ident)?) => {
70   | |         pub trait $visitor_trait_name<'tcx> {
71   | |             // Override these, and call `self.super_xxx` to revert back to the
390  | |                             &mut place.as_ref(),
390  | |                             &mut place.as_ref(),
     | |                             ^^^^^^^^^^^^^^^^^^^ expected struct `mir::PlaceRef`, found `&mir::Place<'_>`
929  | |     }
930  | | }
930  | | }
     | |_- in this expansion of `make_mir_visitor!`
...
1085 |   make_mir_visitor!(MutVisitor, mut);
     |
     |
     = note: expected mutable reference `&mut mir::PlaceRef<'tcx>`
                found mutable reference `&mut &mir::Place<'_>`
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle`
error: could not compile `rustc_middle`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_codegen_ssa" "-p" "rustc_serialize" "-p" "rustc_span" "-p" "rustc_arena" "-p" "rustc_middle" "-p" "rustc_feature" "-p" "rustc_type_ir" "-p" "rustc_query_system" "-p" "rustc_index" "-p" "rustc_hir" "-p" "rustc_apfloat" "-p" "rustc_macros" "-p" "rustc_incremental" "-p" "rustc_graphviz" "-p" "rustc_attr" "-p" "rustc_lexer" "-p" "rustc_ast_pretty" "-p" "rustc_session" "-p" "rustc_lint_defs" "-p" "rustc_target" "-p" "rustc_ast" "-p" "rustc_fs_util" "-p" "rustc_data_structures" "-p" "rustc_symbol_mangling" "-p" "rustc_errors" "-p" "rustc_driver" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_ast_passes" "-p" "rustc_plugin_impl" "-p" "rustc_parse" "-p" "rustc_hir_pretty" "-p" "rustc_error_codes" "-p" "rustc_save_analysis" "-p" "rustc_lint" "-p" "rustc_trait_selection" "-p" "rustc_infer" "-p" "rustc_parse_format" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_interface" "-p" "rustc_builtin_macros" "-p" "rustc_mir_build" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_typeck" "-p" "rustc_ty_utils" "-p" "rustc_passes" "-p" "rustc_privacy" "-p" "rustc_resolve" "-p" "rustc_ast_lowering" "-p" "rustc_traits" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:22
```

</details>

RalfJung commented:
@henryboisdequin thanks! I'll take a look, but I can't promise I'll do that before next weekend. In the mean time, you could also ask on the Rust Zulip; there might be people there that can help you.

henryboisdequin commented:
@RalfJung on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Fixing.20errors.20in.20.2382091/near/226402302) I and a couple of others discussed that it would be impossible to replace `&mut Place` with `&mut PlaceRef` in the mutable visitor. To make such a change, we would require a lot more refactoring which I think should be done in another PR. In this PR, it just addresses your [comment](https://github.com/rust-lang/rust/pull/80865/files#r558978715). I'll start working on the other PR!

henryboisdequin commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

henryboisdequin commented:
@RalfJung `visit_projection` now takes a `PlaceRef`

RalfJung commented on compiler/rustc_middle/src/mir/visit.rs:
```diff
@@ -998,12 +998,11 @@ macro_rules! visit_place_fns {
             location: Location,
         ) {
-            self.super_projection(local, projection, context, location);
+            self.super_projection(place_ref.local, place_ref.projection, context, location);
```
This looks like `super_projection` should get the same treatment, can you also do that? The `super_` and `visit_` methods should be consistent with each other.

RalfJung commented on compiler/rustc_middle/src/mir/visit.rs:
```diff
@@ -1037,15 +1037,14 @@ macro_rules! visit_place_fns {
         ) {
-            let mut cursor = projection;
+            let mut cursor = place_ref.projection;
             while let &[ref proj_base @ .., elem] = cursor {
```
Can't this use `Place::iter_projections` (reversed)? Oh, that's on a place, not a `PlaceRef`... rather silly indeed.
Let's leave that for a future PR, but could you add a FIXME here, like
```
// FIXME: Use PlaceRef::iter_projections, once that exists.
```

RalfJung replied:
(At that point, `visit_projection_elem` should probably also be changed to use `PlaceRef`. But let's first land the work you did so far. :)

RalfJung commented:
Awesome, thanks a lot. :)
Let's see if CI passes, then I'll r+.

RalfJung commented:
@bors r+

bors commented:
:pushpin: Commit a9c61888896da6feea45875813b894e7fa4067f6 has been approved by `RalfJung`

<!-- @bors r=RalfJung a9c61888896da6feea45875813b894e7fa4067f6 -->
<!-- homu: {"type":"Approved","sha":"a9c61888896da6feea45875813b894e7fa4067f6","approver":"RalfJung"} -->

camelid commented:
>  Associated issue: #80647

Should this PR close that issue, or no?

henryboisdequin commented:
> >  Associated issue: #80647
> 
> 
> 
> Should this PR close that issue, or no?

@RalfJung What do you think?

RalfJung commented:
No it should not close the issue, there's still more cleanup left to be done.
